### PR TITLE
Fix legacy widgets not previewing and widgets saving issue

### PR DESCRIPTION
**Pull request by kevin940726**
## Description
Continued and fix #28379.

`useBlockSync` internally uses `cloneBlock` to sync external changes. However, in a recent change in #28379, `cloneBlock` now will sanitize the attributes, which makes `__internalWidgetId` to be dropped off whenever the user is making any changes.

This PR renames `cloneBlock` to `__experimentalCloneSanitizedBlock` (naming TBD) and uses it when duplicating blocks. Then, creates a new method `cloneBlock` that doesn't sanitize attributes, which is essentially the same as the original `cloneBlock` before #28379. This means that the exported `cloneBlock` doesn't have to be breaking anymore.

This also fixes an issue where legacy widgets' previews are not being shown correctly, as `__internalWidgetId` was not being passed to the block before.

## How has this been tested?

Updated e2e test for duplicating blocks, and added e2e test for displaying legacy widgets' previews.

## Types of changes

Breaking change

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

File: packages/blocks/src/api/factory.js [hunk 1]
```diff
@@ -88,8 +88,38 @@ export function createBlocksFromInnerBlocksTemplate(
...
+ * Given a block object, returns a copy of the block object without sanitizing its attributes,
+ * optionally merging new attributes and/or replacing its inner blocks.
+ *
+ * @param {Object} block              Block instance.
+ * @param {Object} mergeAttributes    Block attributes.
+ * @param {?Array} newInnerBlocks     Nested blocks.
+ *
+ * @return {Object} A cloned block.
+ */
+export function duplicateBlock( block, mergeAttributes = {}, newInnerBlocks ) {
```
Comment (kevin940726) [inline packages/blocks/src/api/factory.js:100]: What should be a better name for this? `cloneBlockWithoutSanitizingAttributes`?

Also, is it possible to mark this method private only for now? Should we?

Comment (talldan) [reply packages/blocks/src/api/factory.js:100]: Making it `__experimental` (not for public use and liable to change) is the closest equivalent.

Alternatively there's `__unstable` which is used for anything temporary.

As an aside, I'm not sure sanitize is the right word, as to me this makes it sound like the values of the attributes will be cleaned. But maybe it's too difficult to change that now. (it'd have to be patched into WordPress 5.7 before release, and deprecated in the plugin)

I'd maybe consider something like `cloneWithExtraneousAttributes`, and the sanitize function could be `omitExtraneousAttributes` or `withoutExtraneousAttributes`.

Comment (talldan) [inline packages/blocks/src/api/factory.js:100]: A complete alternative would be a separate pure clone function that accepts the `blockType`, and the widget editor could then declare when the __widgetId attribute should be allowed:
```js
const blockTypeWithWidgetId = withWidgetId( widgetBlockType );
const cloned = cloneBlockUsingBlockType( blockTypeWithWidgetId, widgetBlock );
```

**Review by youknowriad — COMMENTED** [review]
This makes sense to me

**Comment by draganescu** [comment]
It seems this fixes the preview, as it worked predictably now. However I am still able to easily break the widgets editor:

- adding site origin call to action and customizing it
- preview site origin call to action
- adding archives
- save
- reload
- changing some archives property (show post counts)
- changing the test in site origin call to action
- save 
- reload
- the archives block says "Error loading block: Invalid parameter(s): attributes"
- there are two bogus widgets in inactive area (both seem to be a broken archive block)

**Review by talldan — APPROVED** [review]
It makes sense to avoid the breaking change to the API in the upcoming version of WordPress.

I went over the uses of `cloneBlocks` and the changes in this PR look correct:
- Merging blocks - uses `cloneBlocks` to retain the widget id
- useBlockSync - uses `cloneBlocks` to retain the widget id
- Duplicate blocks - sanitizes so that the duplicate doesn't have the widget id
- Patterns - widgetId won't be defined, so `cloneBlocks` is fine.
- Block styles - we don't really care, `cloneBlocks` is fine.

Thanks for fixing @kevin940726, we can separately look at @draganescu's findings.

**Pull request merged**